// ionic-angular/es2015/components/card/card.js
import { Directive, ElementRef, Input, Renderer } from '@angular/core';
import { Config } from '../../config/config';
import { Ion } from '../ion';
export class Card extends Ion {
    constructor(config, elementRef, renderer) {
        super(config, elementRef, renderer, 'card');
    }
    set color(val) {
        this._setColor(val);
    }
    set mode(val) {
        this._setMode(val);
    }
}
Card.decorators = [
    { type: Directive, args: [{
                selector: 'ion-card'
            },] },
];
Card.ctorParameters = [
    { type: Config, },
    { type: ElementRef, },
    { type: Renderer, },
];
Card.propDecorators = {
    'color': [{ type: Input },],
    'mode': [{ type: Input },],
};
export class CardContent extends Ion {
    constructor(config, elementRef, renderer) {
        super(config, elementRef, renderer, 'card-content');
    }
    set color(val) {
        this._setColor(val);
    }
    set mode(val) {
        this._setMode(val);
    }
}
CardContent.decorators = [
    { type: Directive, args: [{
                selector: 'ion-card-content'
            },] },
];
CardContent.ctorParameters = [
    { type: Config, },
    { type: ElementRef, },
    { type: Renderer, },
];
CardContent.propDecorators = {
    'color': [{ type: Input },],
    'mode': [{ type: Input },],
};
export class CardHeader extends Ion {
    constructor(config, elementRef, renderer) {
        super(config, elementRef, renderer, 'card-header');
    }
    set color(val) {
        this._setColor(val);
    }
    set mode(val) {
        this._setMode(val);
    }
}
CardHeader.decorators = [
    { type: Directive, args: [{
                selector: 'ion-card-header'
            },] },
];
CardHeader.ctorParameters = [
    { type: Config, },
    { type: ElementRef, },
    { type: Renderer, },
];
CardHeader.propDecorators = {
    'color': [{ type: Input },],
    'mode': [{ type: Input },],
};
export class CardTitle extends Ion {
    constructor(config, elementRef, renderer) {
        super(config, elementRef, renderer, 'card-title');
    }
    set color(val) {
        this._setColor(val);
    }
    set mode(val) {
        this._setMode(val);
    }
}
CardTitle.decorators = [
    { type: Directive, args: [{
                selector: 'ion-card-title'
            },] },
];
CardTitle.ctorParameters = [
    { type: Config, },
    { type: ElementRef, },
    { type: Renderer, },
];
CardTitle.propDecorators = {
    'color': [{ type: Input },],
    'mode': [{ type: Input },],
};
